### PR TITLE
Expand spawn reason API

### PR DESCRIPTION
**Pull request by Lulu13022002**
Some entity must spawn using some specific method. This allow to change the spawn reason for theses entity as well

### CODE REVIEW

File: patches/api/0203-Spawn-Reason-API.patch [hunk 1]
```diff
@@ -22,33 +22,130 @@ index a89fff5c164881be0286ec2240e94dd5883ecc40..e55f6e2baf35dbd91c433ab9e62713ea
...
+index 352a20da402b3b0182b7a0c69d94397f42054053..6a9604404ed07843c9da91a929651c0d1f1e9d6d 100644
+--- a/src/main/java/org/bukkit/World.java
++++ b/src/main/java/org/bukkit/World.java
+@@ -706,7 +706,27 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
+      * @return Arrow entity spawned as a result of this method
+      */
+     @NotNull
+-    public <T extends AbstractArrow> T spawnArrow(@NotNull Location location, @NotNull Vector direction, float speed, float spread, @NotNull Class<T> clazz);
++    // Paper start
++    default  <T extends AbstractArrow> T spawnArrow(@NotNull Location location, @NotNull Vector direction, float speed, float spread, @NotNull Class<T> clazz) {
```
Comment (lynxplay) [inline patches/api/0203-Spawn-Reason-API.patch]: *absolute nitpick* but there are two spaces between `default` and `<T extends AbstractArrow>` :sweat_smile:

File: patches/api/0203-Spawn-Reason-API.patch [hunk 2]
```diff
@@ -22,33 +22,130 @@ index a89fff5c164881be0286ec2240e94dd5883ecc40..e55f6e2baf35dbd91c433ab9e62713ea
...
++        return spawnArrow(location, direction, speed, spread, clazz, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT); // should be custom but bukkit do that
```
Comment (lynxplay) [inline patches/api/0203-Spawn-Reason-API.patch]: Can you clarify the comment here to "should be custom, but upstream spawns arrows from #spawnArrow with the SpawnReason.DEFAULT" or something along those lines so people know why we use DEFAULT here please.

File: patches/api/0203-Spawn-Reason-API.patch [hunk 3]
```diff
@@ -22,33 +22,130 @@ index a89fff5c164881be0286ec2240e94dd5883ecc40..e55f6e2baf35dbd91c433ab9e62713ea
...
++    /**
++     * Strikes lightning at the given {@link Location}
++     *
++     * @param loc The location to strike lightning
++     * @param doDamage Weather or not this lightning should do damage
++     * @param cause The cause of this lightning
++     * @return The lightning entity.
++     */
++    @NotNull
++    LightningStrike strikeLightning(@NotNull Location loc, boolean doDamage, @NotNull org.bukkit.event.weather.LightningStrikeEvent.Cause cause);
```
Comment (lynxplay) [inline patches/api/0203-Spawn-Reason-API.patch]: I am unsure about the general "doDamage" parameter, That is what `strikeLightningEffect` is there for.
I see how a single unified method might be nicer, but those methods have existed for ages.

**Review by lynxplay — COMMENTED** [review]
Generally looks good, just a few points need discussion before approval.

File: patches/api/0203-Spawn-Reason-API.patch (hunk 2 above)
Comment (Lulu13022002) [reply patches/api/0203-Spawn-Reason-API.patch]: You've right this is more accurate

File: patches/api/0203-Spawn-Reason-API.patch (hunk 3 above)
Comment (Lulu13022002) [reply patches/api/0203-Spawn-Reason-API.patch]: Yeah i will revert, it was just obvious for me

File: patches/api/0203-Spawn-Reason-API.patch (hunk 1 above)
Comment (Lulu13022002) [reply patches/api/0203-Spawn-Reason-API.patch]: ..holy, you should really see that comment: https://github.com/PaperMC/Paper/commit/c6d737528ddb14e360471b39eb97f65e959f88a2#diff-31aba21b5bf05c168ca05351018164e05fc3b26f84ce9d4ac2e19c0c0bbdbf33R22 :eyes:  

But i will fix that, thanks :heart:.

Comment (lynxplay) [reply patches/api/0203-Spawn-Reason-API.patch]: Yea I already left my comment on the merged commit in regards to that PR. Really just a) if you are gonna work here anyway, you can just fix it XD

File: patches/api/0203-Spawn-Reason-API.patch [hunk 4]
```diff
@@ -22,33 +22,141 @@ index a89fff5c164881be0286ec2240e94dd5883ecc40..e55f6e2baf35dbd91c433ab9e62713ea
...
++     */
++    @NotNull
++    LightningStrike strikeLightning(@NotNull Location loc, @NotNull org.bukkit.event.weather.LightningStrikeEvent.Cause cause);
++    // Paper end
+ 
+     /**
+      * Strikes lightning at the given {@link Location} without doing damage
+      *
+      * @param loc The location to strike lightning
++     * @param cause The cause of this lightning
```
Comment (lynxplay) [inline patches/api/0203-Spawn-Reason-API.patch]: We can simply the diff here by following the same layout as the other methods added, e.g. define the ```java
LightningStrike strikeLightningEffect(@NotNull Location loc, @NotNull org.bukkit.event.weather.LightningStrikeEvent.Cause cause);
``` method below, not above.

That way we don't have a weird diff on the javadoc and copy the old methods javadocs into our patch.

**Review by lynxplay — APPROVED** [review]
Beyond my comment, this generally looks good to me :+1:

File: patches/api/0203-Spawn-Reason-API.patch (hunk 4 above)
Comment (Lulu13022002) [reply patches/api/0203-Spawn-Reason-API.patch]: Not sure if i have understand that correctly but i have followed the previous method diff

**Comment by Machine-Maker** [comment]
Can the impl for those new methods be put into a previous patch relating to spawn reasons?

File: patches/api/0197-Spawn-Reason-API.patch [hunk 5]
```diff
@@ -5,50 +5,150 @@ Subject: [PATCH] Spawn Reason API
...
+      * Strikes lightning at the given {@link Location} without doing damage
+      *
+      * @param loc The location to strike lightning
++     * @param cause The cause of this lightning
+      * @return The lightning entity.
+      */
+     @NotNull
+-    public LightningStrike strikeLightningEffect(@NotNull Location loc);
++    LightningStrike strikeLightningEffect(@NotNull Location loc, @NotNull org.bukkit.event.weather.LightningStrikeEvent.Cause cause);
++    // Paper end
```
Comment (Machine-Maker) [inline patches/api/0197-Spawn-Reason-API.patch]: End the paper comment block up top with a single inline comment on the last line.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

File: patches/api/0197-Spawn-Reason-API.patch (hunk 5 above)
Comment (Lulu13022002) [reply patches/api/0197-Spawn-Reason-API.patch]: Idk why but the diff has changed during the rebasing.

**Review by Machine-Maker — APPROVED** [review]

**Comment by Lulu13022002** [comment]
Rebased

**Pull request closed without merging**